otwcode/otwarchive
# AO3-5907 Move share work modal loading to a controller action

# Pull Request Checklist

  ^ Technically the functionality is the same as before but there wasn't a test for anonymous usage of this feature so I added one.

## Issue

https://otwarchive.atlassian.net/browse/AO3-5907

## Purpose

This PR moves loading of "share work" modal into a separate action on the WorksController so that rendering of the share options can happen when users press the share button rather than when they view a work. Part of the motivation for this change is to limit the exposure of user behavior to external sites like twitter and tumblr by not sending http requests to them on every work page load.

Note: In order to accomplish the privacy goal, I moved loading of the twitter widgets javascript into the new share action's view as well. This means that the js would load every time a user shares rather than once per page, but it's the simplest way to ensure that it's _not_ loaded every time the page renders (and my guess would be that people usually only share things once per page load?)

Potential future work could convert bookmark sharing to use a similar asynchronous pattern.

## Testing Instructions

1. View a work that can be shared as a logged-in user with javascript enabled and verify that the "Share" button works.
2. View a work that can be shared as a logged-out user with javascript enabled and verify that the "Share" button works.
3. View a work that can be shared with javascript disabled and verify that the "Share" button is not displayed.
4. Verify in Developer Tools > Network Traffic that loading a work detail page does not cause any http requests to twitter or tumblr.
5. Verify that visiting the share URL (/works/:id/share) for a work in an unrevealed collection does not give away any information about the work.
6. View a work that can be shared as a logged-out user, open the share modal, close the share modal, and log in. Verify that after login you are viewing the work.
7. View a multi-chapter work and verify that the "share" button works

## Credit

*What name do you want us to use to credit your work in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?*

Stephen Burrows

*What pronouns do you prefer (she/her, he/him, zie/hir etc)?*

he/him
